### DISCUSSION TITLE
Squirrel autoformatter

### DISCUSSION THREAD

**Discussion by GeckoEidechse** [R2Northstar/Northstar#681] [Research]
## What?

This has been on my mind since like the early days of Northstar but I realise I never wrote it down.

Titanfall2 uses a custom flavour of [Squirrel](http://www.squirrel-lang.org/) referred to as `squirrel_re` which adds some additional typing functionality among other things.

We should develop (or use an existing tool) to format `squirrel_re` code so that we could have consistent formatting. Such a tool would ideally run both on command line (so that we can use it in CI) and inside some IDE/editor (preferably vscode as it's the most common editor) so that formatting can easily be applied by the developer while they are writing their code.

## Why?

Having an autoformatter for `squirrel_re` could help us tremendously in cleaning up existing code formatting (so many random trailing whitespaces and inconsistent formatting ong) and could also allows us to better compare our changes against vanilla code (by running both vanilla scripts and Northstar modified code through the same formatter and comparing the diff).

It would also automate some of the reviewing work done for NorthstarMods as right now we have to do all formatting checks by hand and some whitespace changes can still slip through.

## Related projects:

By @cpdt
- https://github.com/cpdt/sqparse a Squirrel parser, could be extended to become a formatter ([some work was done on an older version](https://github.com/cpdt/sqparse/tree/sqfmt) but got abandoned)

By @uniboi 
- https://github.com/uniboi/sqr_lsp a Squirrel lsp (state completion unknown)
- https://github.com/uniboi/sqdocs Squirrel docs generator (state completion unknown)
- https://github.com/uniboi/refracture tree sitter based rsquirrel formatter test (incomplete?)
- https://github.com/uniboi/tree-sitter-rsquirrel 

By @r-ex
- https://github.com/r-ex/vscode-squirrel-re

Other
- https://bitbucket.org/marcinbar91/vscode-squirrel/ (Standard Squirrel, not `squirrel_re`)
    https://marketplace.visualstudio.com/items?itemName=marcinbar.vscode-squirrel
- https://bitbucket.org/marcinbar91/vscode-squirrel_linter/ (Standard Squirrel, not `squirrel_re`)
    https://marketplace.visualstudio.com/items?itemName=marcinbar.vscode-squirrel-linter
- https://github.com/harmonytf/vscode-squirrel-linter Linter, Titanfall1 only

# Bounty:

I'm putting a 50€ on this. To be precise, if someone delivers a Squirrel formatter that can be run in CLI and/or CI or some other automated fashion, they will get the bounty payout. I'm happy to increase the payout for additional functionality like e.g. a formatter and parser in one single tool (bonus points for LSP).
For the payout it does not matter whether your tool is written from scratch or forked from one of the existing tools. You'll get the full payout in either case. What counts is the working end result.

**Comment by p0358** [comment]
If you want to implement some state-of-the-art thing for that, then I also suggest adding a support for r1-syntax-only mode, as technically it's for the most part a subset of r2 one, so on paper it shouldn't be that much more effort if already doing it all, right.......?

**Comment by Bobbyperson** [comment]
https://github.com/Bobbyperson/sqformat

Working on this
